### ISSUE TITLE
Improve advice / diagnostics when VI gives terrible results

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pymc#6271]
I have never used VI before, so I don't know if it's expected to fail so easily:

```python
import pymc as pm

with pm.Model():
    x = pm.Normal("x")
    # increasing the distance between prior and posterior makes no difference for VI
    # e.g., sigma=1e-3 or observed=100
    y = pm.Normal("y", x, sigma=0.5, observed=10)  
    
    nuts = pm.sample()
    vi = pm.fit()
    
print((
    nuts.posterior["x"].mean().values, 
    vi.sample().posterior["x"].mean().values
))  # (array(8.01670159), array(3.11480314))
```

**Comment by ricardoV94** [comment]
> which is not a great estimate from either method!

The example is conditioning on a single observation. With a looser likelihood it's normal that the posterior gets closer to the prior.

**Comment by fonnesbeck** [comment]
Ah, you're right. I was confusing the likelihood and the prior. 

Yes, this is odd behavior. FWIW, I get pretty reasonable results on my bigger models (at least as far as point estimates go).

**Comment by ricardoV94** [comment]
V3 gives the same results...

**Comment by ColCarroll** [comment]
Just so we're on the same page, I think the correct posterior here is 
```
x | y ~ N(obs / (1 + sigma^2), sigma^2 / (1 + sigma^2))
```

so the most recent example with `obs = 100`, `sigma = 1e-5` should be
```
N(99.99999999,  0.9999e-11)
```
and the first example with `obs=10`, `sigma=0.5` should be
```
N(8, 0.2)
```

**Comment by fonnesbeck** [comment]
VI should not have a problem approximating this well. Any thoughts, @ferrine ?

**Comment by ferrine** [comment]
The learning rate seems to be too small for this model. This code helps to investigate

```python
with model:
    advi = pm.ADVI()
    tracker = pm.callbacks.Tracker(
        mean=advi.approx.mean.eval,  # callable that returns mean
        std=advi.approx.std.eval,  # callable that returns std
    )
    approx = advi.fit(callbacks=[tracker])
fig = plt.figure(figsize=(16, 9))
mu_ax = fig.add_subplot(221)
std_ax = fig.add_subplot(222)
hist_ax = fig.add_subplot(212)
mu_ax.plot(tracker["mean"])
mu_ax.set_title("Mean track")
std_ax.plot(tracker["std"])
std_ax.set_title("Std track")
hist_ax.plot(advi.hist)
hist_ax.set_title("Negative ELBO track");
```
![image](https://user-images.githubusercontent.com/11705326/202126088-37728dea-9815-4796-be34-fae94cc5344a.png)

When I increase learning rate to be 1e-2:
```python
approx = advi.fit(callbacks=[tracker], obj_optimizer=pm.adagrad_window(learning_rate=1e-2))
```
I get a different picture
![image](https://user-images.githubusercontent.com/11705326/202126460-05641401-e2da-4d9d-8261-54a113c7db85.png)

**Comment by aseyboldt** [comment]
Maybe the tracker could even be enabled by default, and the results stored in the approx, just as with the sample stats for mcmc? It sounds like checking the convergence of the optimizer should be standard the same way we do this with MCMC.

**Comment by ricardoV94** [comment]
Does VI think optimization has plateaued? Or does it have a max number of iterations?

**Comment by fonnesbeck** [comment]
@ricardoV94 10K is the default number of iterations

**Comment by ricardoV94** [comment]
Oh I was confusing with find_MAP where it interrupts once it stops improving. VI always goes the 10k draws?

**Comment by cfonnesbeck** [comment]
Just like sampling, its whatever fixed number of iterations that you pass it. For my production models at work, I typically use 50-100K. There is no automated early stopping, though that would be a nice feature.

**Comment by aseyboldt** [comment]
Automatic diagnostics would be great, but I think even visual checks like the plots above are a good start. My point was that they should be _very_ easy to access, and tutorials should show them as well I think.
